src/foundation/components/InfiniteMessageList/index.tsx

```diff
@@ -0,0 +1,142 @@
+  const {
+    messages,
+    renderMessage,
+
+    messageTopArea,
+    messageBottomArea,
+    overlayArea,
+
+    onLoadPrev,
+    onLoadNext,
+    loadThreshold = 0.05,
+
+    onScrollPosition = noop,
+    depsForResetScrollPositionToBottom,
+
+    scrollPositionRef: _scrollPositionRef,
+    scrollDistanceFromBottomRef: _scrollDistanceFromBottomRef,
+  } = props;
+
+  const listRef = ref && 'current' in ref ? ref : { current: null };
+  const isFetching = useRef(false);
+  const direction = useRef<'top' | 'bottom'>();
+  const oldScrollTop = useRef(0);
+
+  const internal_scrollPositionRef = useRef(0);
+  const internal_scrollDistanceFromBottomRef = useRef(0);
+  const scrollPositionRef = _scrollPositionRef ?? internal_scrollPositionRef;
+  const scrollDistanceFromBottomRef = _scrollDistanceFromBottomRef ?? internal_scrollDistanceFromBottomRef;
+
+  // SideEffect: scroll to bottom on initialized
+  useLayoutEffect(() => {
+    if (listRef.current) {
+      listRef.current.scrollTop = listRef.current.scrollHeight;
+    }
+  }, depsForResetScrollPositionToBottom);
+
+  // SideEffect: keep scroll position
+  useLayoutEffect(() => {
+    if (listRef.current) {
+      if (direction.current === 'top') {
+        listRef.current.scrollTop = listRef.current.scrollHeight - scrollPositionRef.current;
+      }
+      if (direction.current === 'bottom') {
+        listRef.current.scrollTop = oldScrollTop.current;
+      }
+      direction.current = undefined;
+    }
+  }, [listRef.current, messages.length]);
+
+  const handleScroll: UIEventHandler<HTMLDivElement> = async () => {
+    if (!listRef.current) return;
+    const list = listRef.current;
+
+    onScrollPosition(getReachedStatus(list));
+
+    scrollPositionRef.current = list.scrollHeight - list.scrollTop;
+    scrollDistanceFromBottomRef.current = scrollPositionRef.current - list.clientHeight;
+    oldScrollTop.current = list.scrollTop;
+
+    if (isFetching.current) return;
+
+    const threshold = list.clientHeight * Math.min(Math.max(0, loadThreshold), 1);
+    if (list.scrollTop <= threshold) {
+      isFetching.current = true;
+      direction.current = 'top';
+      await onLoadPrev();
+      isFetching.current = false;
+    } else if (list.scrollHeight - list.scrollTop - list.clientHeight <= threshold) {
+      isFetching.current = true;
+      direction.current = 'bottom';
+      await onLoadNext();
+      isFetching.current = false;
+    } else {
+      direction.current = undefined;
+    }
+  };
+
+  return (
+    <div className="sendbird-conversation__scroll-container">
+      <div className="sendbird-conversation__padding" />
+      <div
+        ref={listRef}
+        className="sendbird-conversation__messages-padding"
+        data-testid="sendbird-message-list-container"
+        onScroll={handleScroll}
+      >
+        {messageTopArea}
+        {messages.map((message, index) => renderMessage({ message, index }))}
+        {messageBottomArea}
```

Comment:
any idea how to add welcome messages?
also how about typing indicator?

Reply:
The prop names are currently set to welcomeMessages and typingIndicator, but they have been changed to more generic names. Any component can be placed in these positions, so it can include welcome messages and typing indicators.

```tsx
<InfiniteMessageList
  messageTopArea={<WelcomeMessages />}
  messageBottom={<TypingIndicator />}
/>
```